### PR TITLE
File diff should compare against "merge base" not current state of the PR's base branch

### LINKED ISSUE #1007: Discrepancy between github.com and GHfVS PR diffs
- GitHub Extension for Visual Studio version: 2.2.1.100
- Visual Studio version: 2015

I checked out a PR and noticed that some PR files seemed have no changes:

image

On github.com the same file does appear to have changes:

image

### PR DESCRIPTION
**Pull request by jcansdale**
The PR's base branch might well have moved on since the PR was originally created. The PR head should be compared against the merge base between the PR base and head.

1. Check if merge base can be found between base and head SHAs.
2. If it can, diff using local merge base and PR head files.
3. If it can't, fetch the PR's base and head branches.
4. Find merge base and diff against PR head file.

This means at most one fetch is required before the PR files can be explored without hitting the network again. 😄 

Fixes #1007

### CODE REVIEW

File: src/GitHub.App/Services/GitClient.cs [hunk 1]
```diff
@@ -14,6 +14,7 @@ namespace GitHub.Services
 {
     [Export(typeof(IGitClient))]
     [PartCreationPolicy(CreationPolicy.Shared)]
+    [NullGuard(ValidationFlags.None)]
```
Comment (jcansdale) [inline src/GitHub.App/Services/GitClient.cs]: Not sure if I still need this.

Comment (grokys) [inline src/GitHub.App/Services/GitClient.cs]: Leave it on there, `NullGuard` being enabled causes more problems than it solves IMO and we will hopefully remove it soon.

File: src/UnitTests/GitHub.App/Services/PullRequestServiceTests.cs [hunk 2]
```diff
@@ -26,27 +26,72 @@ public class TheExtractDiffFilesMethod
...
+            var headSha = "headSha";
+            var mergeBaseSha = null as string;
+            var checkedOut = false;
+            var expectMessage = $"Couldn't find merge base between {baseSha} and {headSha}.";
+
+            FileNotFoundException ex = null;
+            try
+            {
+                // Can't use Assert.Throws because ExtractDiffFiles is async.
+                await ExtractDiffFiles(baseSha, baseFileContent, headSha, headFileContent, mergeBaseSha, mergeBaseFileContent,
```
Comment (grokys) [inline src/UnitTests/GitHub.App/Services/PullRequestServiceTests.cs]: Can't you use `Assert.ThrowsAsync` here?

File: src/UnitTests/GitHub.App/Services/PullRequestServiceTests.cs [hunk 3]
```diff
@@ -79,69 +124,66 @@ public class TheExtractDiffFilesMethod
...
             var headRef = new GitReferenceModel("ref", "label", headSha, "uri");
-            var headFileSha = "headFileSha";
             var checkedOut = false;
 
-            var files = await ExtractDiffFiles(baseFileContent, headFileContent, fileName, baseSha, headSha, headFileSha, checkedOut, modelServiceHeadFile: headFile);
+            FileNotFoundException ex = null;
+            try
+            {
+                // Can't use Assert.Throws because ExtractDiffFiles is async.
+                await ExtractDiffFiles(baseSha, baseFileContent, headSha, headFileContent, baseSha, baseFileContent,
```
Comment (grokys) [inline src/UnitTests/GitHub.App/Services/PullRequestServiceTests.cs]: Likewise here.

Comment (jcansdale) [inline src/UnitTests/GitHub.App/Services/PullRequestServiceTests.cs]: It looks like you can indeed. I hadn't noticed `Assert.ThrowsAsync` before. Thanks for the tip!

**Pull request merged**